### PR TITLE
Locking

### PR DESCRIPTION
**Pull request by rawxfish**
Pessimistic locking for #49

### CODE REVIEW

File: puzzles/views.py [hunk 1]
```diff
@@ -190,24 +197,25 @@ def delete_puzzle(request, pk):
...
         return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/'))
-    puzzle = get_object_or_404(Puzzle, pk=pk)
+    puzzle = get_object_or_404(Puzzle.objects.select_for_update(), pk=pk)
     (tag, _) = PuzzleTag.objects.update_or_create(
         name=form.cleaned_data["name"],
         defaults={'color' : form.cleaned_data["color"]}
     )
     if tag.is_meta:
-        metapuzzle = Puzzle.objects.get(name=tag.name)
+        metapuzzle = get_object_or_404(Puzzle, name=tag.name)
```
Comment (erwa) [inline puzzles/views.py]: Do we need to use `Puzzle.objects.select_for_update()` here as well?

File: puzzles/views.py [hunk 2]
```diff
@@ -79,22 +81,23 @@ def slack_guess(request):
...
     answer_text = __sanitize_guess(slack_message.get('text'))
     channel_id = slack_message.get('channel_id')
-    puzzle = Puzzle.objects.get(channel=channel_id)
+    puzzle = get_object_or_404(Puzzle.objects.select_for_update(), channel=channel_id)
     print("puzzle: " + str(puzzle))
     if puzzle.status == Puzzle.SOLVED:
         return HttpResponse("Puzzle is already solved!")
-    if Answer.objects.filter(puzzle=puzzle).filter(text=answer_text):
+
+    answer, created = Answer.objects.get_or_create(puzzle=puzzle, text=answer_text)
```
Comment (erwa) [inline puzzles/views.py]: Nit: I think you can use `_, created` instead.

File: puzzles/views.py (hunk 1 above)
Comment (rawxfish) [reply puzzles/views.py]: Hmm, I'm not sure. I didn't think we needed it since we aren't updating metapuzzle.

Comment (erwa) [reply puzzles/views.py]: Could an inconsistency arise if the metapuzzle were changed to a regular puzzle while this transaction was happening? Then on line 218, we would be treating the now-regular puzzle as the meta for this puzzle?

Comment (rawxfish) [reply puzzles/views.py]: Yeah you're right, a change to metapuzzle could happen from a third puzzle.

**Pull request merged**